hints/hint_distribution.py

```diff
@@ -0,0 +1,450 @@
+]
+
+
+class InvalidHintDistribution(Exception):
+    pass
+
+
+class HintDistribution:
+    def __init__(self):
+        self.banned_stones = []
+        self.added_locations = []
+        self.removed_locations = []
+        self.added_items = []
+        self.removed_items = []
+        self.sots_dungeon_placed = 0
+        self.dungeon_sots_limit = 0
+        self.sots_dungeon_placed = 0
+        self.dungeon_barren_limit = 0
+        self.distribution = {}
+        self.rng: Random = None
+        self.logic = None
+        self.hints = []
+        self.weighted_types = []
+        self.weights = []
+        self.sots_locations = []
+        self.barren_overworld_zones = []
+        self.placed_ow_barren = 0
+        self.barren_dungeons = []
+        self.placed_dungeon_barren = 0
+        self.prev_barren_type = None
+        self.hinted_locations = []
+        self.weighted_types = []
+        self.weights = []
+        self.hintable_items = []
+        self.junk_hints = []
+        self.sometimes_hints = []
+        self.hint_descriptors = []
+
+    def read_from_file(self, f):
+        self._read_from_json(json.load(f))
+
+    def read_from_str(self, s):
+        self._read_from_json(json.loads(s))
+
+    def _read_from_json(self, jsn):
+        self.banned_stones = jsn["banned_stones"]
+        self.added_locations = jsn["added_locations"]
+        self.removed_locations = jsn["removed_locations"]
+        self.added_items = jsn["added_items"]
+        self.removed_items = jsn["removed_items"]
+        self.dungeon_sots_limit = jsn["dungeon_sots_limit"]
+        self.dungeon_barren_limit = jsn["dungeon_barren_limit"]
+        self.distribution = jsn["distribution"]
+
+    """
+    Performs initial calculations and populates the distributions internal
+    tracking mechanisms for hint generation
+    """
+
+    def start(self, logic: Logic, always_hints: list, sometimes_hints: list):
+        self.rng = logic.rando.rng
+        self.logic = logic
+
+        hint_descriptors = {}
+        for name, loc in self.logic.item_locations.items():
+            if "text" in loc:
+                hint_descriptors[name] = loc["text"]
+            else:
+                hint_descriptors[name] = name + " has"
+        self.hint_descriptors = hint_descriptors
+
+        for loc in self.added_locations:
+            location = loc["location"]
+            if loc["type"] == "always":
+                if location in always_hints:
+                    continue
+                always_hints.append(location)
+                if location in sometimes_hints:
+                    if location in sometimes_hints:
+                        continue
+                    sometimes_hints.remove(location)
+            elif loc["type"] == "sometimes":
+                sometimes_hints.append(location)
+                if location in sometimes_hints:
+                    always_hints.remove(location)
```

Comment:
```suggestion
                if location not in sometimes_hints:
                    sometimes_hints.append(location)
```
not sure if that's what you want to do here